File: data-structures-and-algorithms/graphs/medium/possible-bipartition/dfs/possible_bipartition.py
class Solution(object):
    """
    Return `True` or `False` whether a given set of `N` people can be split into
    two groups of any size while keeping people in the given `dislikes` adjacency
    list in separate groups.
    """
    def possibleBipartition(self, N, dislikes):
        graph = collections.defaultdict(list)

        for u, v in dislikes:
            graph[u].append(v)
            graph[v].append(u)

        color = {}

        def dfs(node, c = 0):
            if node in color:
                return color[node] == c

            color[node] = c

            return all(dfs(neighbor, c ^ 1) for neighbor in graph[node])

        return all(dfs(node)
            for node in range(1, N+1)
            if node not in color)
          

File: data-structures-and-algorithms/graphs/medium/possible-bipartition/dfs/possibleBipartition.js
/**
 * @param {number[]} N 
 * @param {number[][]} dislikes 
 * @return {boolean}
 */
function possibleBipartition(N, dislikes) {
  if (!dislikes.length) return true

  let colors = Array(N + 1).fill(0)
  let graph = {}

  for (let [ a, b ] of dislikes) {
    graph[ a ] = (graph[ a ] || new Set()).add(b)
    graph[ b ] = (graph[ b ] || new Set()).add(a)
  }

  for (let i = 1; i <= N; i++) {
    if (!colors[ i ] && !dfs(i, 1)) return false
  }

  function dfs(node, color) {
    if (colors[ node ]) return colors[ node ] == color
    if (!graph[ node ]) return true

    colors[ node ] = color

    for (let edge of graph[ node ]) {
      if (!dfs(edge, ~color)) return false
    }

    return true
  }

  return true
}